### PR TITLE
tables: Add constraints and testing

### PR DESCRIPTION
**Pull request by directionless**
After https://github.com/osquery/osquery/pull/6103 I thought I'd take a pass at the spec files and tests. Not totally sure on all of these.

Also adds an `IntOrEmpty` type to the validation map

Requires: https://github.com/osquery/osquery/pull/6104

Relates To: https://github.com/osquery/osquery/pull/5932

### CODE REVIEW

File: specs/carves.table [hunk 1]
```diff
@@ -1,13 +1,13 @@
 table_name("carves")
 description("Forensic Carves.")
 schema([
-    Column("time", BIGINT, "Time at which the carve was kicked off"),
```
Comment (theopolis) [inline specs/carves.table]: Why are these all set to `additional`?

I only see `path` and `carve` being used in the table implementation.

File: specs/darwin/running_apps.table [hunk 2]
```diff
@@ -3,9 +3,9 @@ table_name("running_apps")
 description("OSX applications currently running on the host system.")
 
 schema([
-    Column("pid", INTEGER, "The pid of the application"),
-    Column("bundle_identifier", TEXT, "The bundle identifier of the application"),
-    Column("is_active", INTEGER, "1 if the application is in focus, 0 otherwise")
+    Column("pid", INTEGER, "The pid of the application", additional=True),
+    Column("bundle_identifier", TEXT, "The bundle identifier of the application", additional=True),
+    Column("is_active", INTEGER, "1 if the application is in focus, 0 otherwise", index=True)
```
Comment (theopolis) [inline specs/darwin/running_apps.table]: This is not correct, `index` indicates a primary key.

File: specs/posix/docker_container_mounts.table [hunk 3]
```diff
@@ -3,7 +3,7 @@ description("Docker container mounts.")
 schema([
     Column("id", TEXT, "Container ID", index=True),
     Column("type", TEXT, "Type of mount (bind, volume)"),
-    Column("name", TEXT, "Optional mount name"),
+    Column("name", TEXT, "Optional mount name", index=True),
```
Comment (theopolis) [inline specs/posix/docker_container_mounts.table]: Is this `name` unique (e.g., a pkey?) or could the same `name` show up in different container IDs?

File: specs/posix/docker_container_networks.table [hunk 4]
```diff
@@ -2,7 +2,7 @@ table_name("docker_container_networks")
 description("Docker container networks.")
 schema([
     Column("id", TEXT, "Container ID", index=True),
-    Column("name", TEXT, "Network name"),
+    Column("name", TEXT, "Network name", index=True),
```
Comment (theopolis) [inline specs/posix/docker_container_networks.table]: Same as above?

File: specs/posix/docker_container_processes.table [hunk 5]
```diff
@@ -3,7 +3,7 @@ description("Docker container processes.")
 schema([
     Column("id", TEXT, "Container ID", index=True, required=True),
     Column("pid", BIGINT, "Process ID"),
-    Column("name", TEXT, "The process path or shorthand argv[0]"),
+    Column("name", TEXT, "The process path or shorthand argv[0]", index=True),
```
Comment (theopolis) [inline specs/posix/docker_container_processes.table]: Ditto

File: tests/integration/tables/asl.cpp [hunk 6]
```diff
@@ -22,32 +22,25 @@ class asl : public testing::Test {
...
+      {"gid", IntType},
+      {"uid", IntType},
+      {"level", IntType},
+      {"message", NormalType},
+      {"ref_pid", IntOrEmpty},
+      {"ref_proc", NormalType},
+      {"extra", NormalType},
+  };
+
+  auto const data = execute_query("select * from asl limit 5");
```
Comment (theopolis) [inline tests/integration/tables/asl.cpp]: Is this a flaky test? Can we assure there are at least 1 rows?

File: tests/integration/tables/asl.cpp [hunk 7]
```diff
@@ -22,32 +22,25 @@ class asl : public testing::Test {
...
+  ASSERT_GT(data.size(), 1ul);
```
Comment (theopolis) [inline tests/integration/tables/asl.cpp]: I recommend `ASSERT_FALSE(data.empty())`

File: tests/integration/tables/crashes.cpp [hunk 8]
```diff
@@ -22,35 +22,30 @@ class crashes : public testing::Test {
...
+      {"crashed_thread", IntType},
+      {"stack_trace", NormalType},
+      {"exception_type", NormalType},
+      {"exception_codes", NormalType},
+      {"exception_notes", NormalType},
+      {"registers", NormalType},
+  };
+
+  auto const data = execute_query(
+      "select crashes.* from users CROSS JOIN crashes "
```
Comment (theopolis) [inline tests/integration/tables/crashes.cpp]: This seems flaky, we cannot assure there is a crash on a machine.

File: tests/integration/tables/helper.cpp [hunk 9]
```diff
@@ -203,6 +203,20 @@ bool is_valid_hex(const std::string& value) {
 }
 
 bool validate_value_using_flags(const std::string& value, int flags) {
+  // Early return on EmptyOk
```
Comment (theopolis) [inline tests/integration/tables/helper.cpp]: Should we assert that `EmptyOk` and `NonEmpty` are not both set?

File: tests/integration/tables/keychain_items.cpp [hunk 10]
```diff
@@ -23,7 +23,7 @@ class KeychainItemsTest : public testing::Test {
 
 TEST_F(KeychainItemsTest, test_sanity) {
   auto const data = execute_query("select * from keychain_items");
-  ASSERT_GT(data.size(), 0ul);
+  ASSERT_GT(data.size(), 1ul);
```
Comment (theopolis) [inline tests/integration/tables/keychain_items.cpp]: Why the change here?

File: tests/integration/tables/ntfs_acl_permissions.cpp [hunk 11]
```diff
@@ -22,24 +22,14 @@ class ntfsAclPermissions : public testing::Test {
...
-  //}
-  // 4. Perform validation
-  // validate_rows(data, row_map);
+  ValidationMap row_map =
+  { {"path", NormalType} {"type", NormalType} {"principal", NormalType} {
+      "access", NormalType} {"inherited_from", NormalType} }
+
+  auto const data =
+      execute_query("select * from ntfs_acl_permissions where path = '\'");
+  ASSERT_GE(data.size(), 1ul);
```
Comment (theopolis) [inline tests/integration/tables/ntfs_acl_permissions.cpp]: Ditto about non-empty.

File: tests/integration/tables/preferences.cpp [hunk 12]
```diff
@@ -22,26 +22,25 @@ class preferences : public testing::Test {
...
-  //}
-  // 4. Perform validation
-  // validate_rows(data, row_map);
+  ASSERT_GE(data.size(), 1ul);
+  validate_rows(data, row_map);
+
+  auto const datajoin = execute_query(
+      "select users.username, preferences.* from users CROSS JOIN preferences "
+      "USING(username) where preferences.domain = 'com.apple.Preferences';");
+  ASSERT_GE(datajoin.size(), 1ul);
```
Comment (theopolis) [inline tests/integration/tables/preferences.cpp]: Ditto non-empty.

File: tests/integration/tables/preferences.cpp [hunk 13]
```diff
@@ -22,26 +22,25 @@ class preferences : public testing::Test {
...
-  //      {"key", NormalType}
-  //      {"subkey", NormalType}
-  //      {"value", NormalType}
-  //      {"forced", IntType}
-  //      {"username", NormalType}
-  //      {"host", NormalType}
-  //}
-  // 4. Perform validation
-  // validate_rows(data, row_map);
+  ASSERT_GE(data.size(), 1ul);
```
Comment (theopolis) [inline tests/integration/tables/preferences.cpp]: Ditto non-empty.

File: tests/integration/tables/process_open_sockets.cpp [hunk 14]
```diff
@@ -22,31 +22,23 @@ class processOpenSockets : public testing::Test {
...
-  //      {"remote_address", NormalType}
-  //      {"local_port", IntType}
-  //      {"remote_port", IntType}
-  //      {"path", NormalType}
-  //      {"state", NormalType}
-  //      {"net_namespace", NormalType}
-  //}
-  // 4. Perform validation
-  // validate_rows(data, row_map);
+  ASSERT_GE(data.size(), 1ul);
```
Comment (theopolis) [inline tests/integration/tables/process_open_sockets.cpp]: Ditto non-empty.

File: specs/posix/docker_container_mounts.table (hunk 3 above)
Comment (directionless) [reply specs/posix/docker_container_mounts.table]: I was super unsure on these. There's a bunch of indirection in these tables

File: tests/integration/tables/keychain_items.cpp (hunk 10 above)
Comment (directionless) [reply tests/integration/tables/keychain_items.cpp]: User error. This should be testing for not-empty, and I'm used to `GE 1ul` and not `GT 0ul`

File: tests/integration/tables/asl.cpp (hunk 6 above)
Comment (directionless) [reply tests/integration/tables/asl.cpp]: No clue! I thought I'd PR this and we'd see how the tests came back

File: tests/integration/tables/asl.cpp (hunk 7 above)
Comment (directionless) [reply tests/integration/tables/asl.cpp]: Huh. That seems reasonable, though I think we use `ASSERT_GT(data.size(), 1ul);` in a lot of places. I'd be happy to batch change them all to `ASSERT_FALSE(data.empty())`

File: tests/integration/tables/crashes.cpp (hunk 8 above)
Comment (directionless) [reply tests/integration/tables/crashes.cpp]: Could be. I think it's worth the joined query, but I agree if it proves flakey we should not require non-empty. As with others, I PR'ed something to see what it'd do

File: tests/integration/tables/helper.cpp (hunk 9 above)
Comment (directionless) [reply tests/integration/tables/helper.cpp]: Not a bad optimization. I didn't think of it

**Comment by theopolis** [comment]
```
85: [----------] 1 test from processOpenSockets
85: [ RUN      ] processOpenSockets.test_sanity
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/helper.cpp:162: Failure
85: Expected: (validation_map.count(rec.first)) != (std::size_t{0}), actual: 0 vs 0
85: Unexpected column "net_namespace" in a row
85: [  FAILED  ] processOpenSockets.test_sanity (10 ms)
85: [----------] 1 test from processOpenSockets (10 ms total)
```

```
85: [----------] 1 test from pythonPackages
85: [ RUN      ] pythonPackages.test_sanity
85: /__w/1/s/usr/src/debug/osquery/tests/integration/tables/python_packages.cpp:36: Failure
85: Value of: data.empty()
85:   Actual: true
85: Expected: false
85: [  FAILED  ] pythonPackages.test_sanity (1 ms)
85: [----------] 1 test from pythonPackages (1 ms total)
```
